core/bentley/src/Assert.ts

```diff
@@ -6,18 +6,56 @@
+ *  1 They allow the programmer to declare conditions that they believe cannot possibly occur. If such conditions occur, they indicate
+ *    a serious flaw in the programmer's logic.
+ *  2 They allow the programmer to assure the TypeScript compiler of the truth of some condition that the compiler cannot itself infer.
+ *  3 They allow the author of an API to indicate to consumers of the API a serious misuse that should be corrected during development.
+ *
+ * Assertions should **never** be used to test for conditions - however unlikely - that could be expected to occur at run-time,
+ * such as failing to write to a file or load a resource over the network. If the condition asserted ever fails in a production environment,
+ * the programmer has made a serious mistake.
+ *
+ * By default, `assert` does nothing unless the environment variable `NODE_ENV` exists and is set to `development`. Use [[setAssertionsEnabled]] to override this behavior.
+ *
+ * Note that even when assertions are disabled, calls to `assert` remain in the code and their arguments will be evaluated at run-time.
+ * Therefore, if your condition or message requires computation, prefer to pass it as a function to prevent it from being evaluated when assertions are disabled.
+ *
+ * @param condition The condition that is asserted to be `true`. If the condition is more complex than a simple `boolean` variable, pass it as a function to prevent it from being evaluated when assertions are disabled.
+ * @param message An optional description of the condition being asserted, to be included in the exception if `condition` is `false`. If the message must be computed, pass it as a function to prevent it from being evaluated when assertions are disabled. Defaults to "Programmer Error".
+ * @throws Error containing the specified `message` if `condition` is `false`.
+ * @see [[setAssertionsEnabled]] to enable or disable assertions.
+ * @public
+ */
+export function assert(condition: boolean | (() => boolean), message?: string | (() => string)): asserts condition {
+  if (!assertionsEnabled)
+    return;
+
+  if ("boolean" !== typeof condition)
+    condition = condition();
+
+  if (condition)
+    return;
+
+  message = message ?? "Programmer Error";
+  if ("string" !== typeof message)
+    message = message();
+
+  throw new Error(`Assert: ${message}`);
+}
+
+/** Sets whether the [[assert]] function is enabled.
+ * When [[assert]] is disabled, it returns immediately without evaluating the asserted condition. By default, it is enabled only if
+ * the environment variable `NODE_ENV` exists and is set to `development`.
+ * @param enabled If true, `assert` will throw upon a false assertion.
+ * @returns true if assertions were enabled prior to this call.
  * @public
  */
-export function assert(condition: boolean, msg?: string): asserts condition {
-  if (!condition)
-    throw new Error(`Assert: ${msg ?? "Programmer Error"}`);
+export function setAssertionsEnabled(enabled: boolean): boolean {
```

Comment:
Is this really necessary?  I just did some testing, and without defining this function, Terser is able to do some really impressive optimizations - basically it completely remove calls to assert without any special configuration.

But as soon as it sees this function definition, it (rightly) determines that it can no longer safely assume all that code can be removed.

Reply:
Here's an example:

```ts
import { setAssertionsEnabled, assert } from "./Assert";
assert(someTest());
```
would currently webpack/minify down to (with whitespace re-added):
```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(o, e, r) {
        o.exports = r(1)
    }, function(o, e, r) {
        "use strict";
        r.r(e);
        let n = !1;
        ! function(o, e) {
            if (n && ("boolean" != typeof o && (o = o()), !o)) throw "string" != typeof(e = e || "Programmer Error") && (e = e()), new Error("Assert: {message}")
        }(someTest())
    }],
    [
        [0, 1]
    ]
]);
```
but as soon as I delete the definition for `setAssertionsEnabled`, that same example bundled becomes:
```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(n, o, s) {
        n.exports = s(1)
    }, function(n, o, s) {
        "use strict";
        s.r(o);
        someTest()
    }],
    [
        [0, 1]
    ]
]);
```

Reply:
Paul, is there a scenario where you'd want to turn off asserts in development?

Reply:
Yeah, @pmconne, just to clarify a bit, for frontend builds, `process.env.NODE_ENV` is usually replaced with a compile-time constant (either `"development"` or `"production"`).  This is a truly bonkers convention, but a fairly well-established one, and most bundling tools support it out of the box.

Reply:
You're going to have to tell me what I'm looking at there, @wgoehrig.
It looks to me like the call to `assert` is being elided at bundle time, which is not what I want. I want assertions enabled or disabled at run-time based on NODE_ENV.

Reply:
@wgoehrig that means that apps who consume are packages will never see assertions from our packages' code. That is not entirely desirable.

Reply:
That's the crucial intermediate step here tho - `process.env.NODE_ENV` becomes `"production"`, so

```
let assertionsEnabled = process.env.NODE_ENV === "development";
```
effectively becomes
```
let assertionsEnabled = false;
```
and Terser takes over from there.  Personally, I was surprised that just removing `setAssertionsEnabled` worked - I would have assumed you'd at least also needed to make `assertionsEnabled` const, but I guess Terser was able to see that `assertionsEnabled` wasn't exported and nothing in the file was left to ever mutate it...

Reply:
Sorry, GH is not refreshing these fast enough and I keep responding to older comments.

@pmconne the convention everyone uses is basically the following:

- `npm start` does a hot-reloading _development_ build, so asserts are still present
- `npm run build` does a production build, which minifies, disables asserts, etc.

Since all this minification only happens when a whole _app_ is bundled, consumers of our packages will still see asserts during development of their app.

Reply:
So @wgoehrig, talk to me like a webpack idiot, because I am: if somebody installs core-frontend from npm and bundles it with their app, they will see assertions from core-frontend as well as from their own code IFF process.env.NODE_ENV is defined when they bundle their app?

Reply:
@wgoehrig yeah, we crossed streams again.
Given that, the only purpose of `setAssertionsEnabled` is tests. The tests were useful for testing these changes but I'm ok with removing them now.

Reply:
Unless we can set NODE_ENV=development when we webpack the tests?

Reply:
Almost.  The value of _your_ process.env.NODE_ENV at build time is not actually relevant - the constant injected into the code comes purely from the [webpack configuration](https://webpack.js.org/configuration/optimization/#optimizationnodeenv).  Again, using process.env in the name there is super misleading and downright bizarre, but that's the established convention, and many of _our_ dependencies require it.

So basically everyone just has two different webpack configs, and they use them for the different npm scripts.

Reply:
> Unless we can set NODE_ENV=development when we webpack the tests?

Yeah, that should be easy enough - I'll look into that.

Reply:
@wgoehrig nah don't worry about it-  most of the tests want to turn assertions on and off anyway.
Just to be clear: what was process.NODE_ENV set to when you produced the terser output?

Reply:
Actually, good news - all our test webpack configs are already setting that.

Reply:
I did not have `process.NODE_ENV` set at all when I produced either of those outputs, but my webpack config had [mode](https://webpack.js.org/configuration/mode/) set to "production".

If I explicitly override the config to set NODE_ENV to "development" (but still optimize/minify - not a usual combination), I get:

```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(o, r, e) {
        o.exports = e(1)
    }, function(o, r, e) {
        "use strict";
        e.r(r);
        ! function(o, r) {
            if ("boolean" != typeof o && (o = o()), !o) throw "string" != typeof(r = r || "Programmer Error") && (r = r()), new Error("Assert: {message}")
        }(someTest())
    }],
    [
        [0, 1]
    ]
]);
"
```

Reply:
@wgoehrig then help me understand. Is `someTest()` being invoked below?
```
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(n, o, s) {
        n.exports = s(1)
    }, function(n, o, s) {
        "use strict";
        s.r(o);
        someTest()
    }],
    [
        [0, 1]
    ]
]);
```

Reply:
Short answer is yes.  Terser didn't know whether `someTest()` had side effects and so that wasn't safe to remove, even when eliding the call to `assert`.

All the examples I gave are incomplete because there's actually another file with the rest of the webpack boilerplate.  Webpack now defaults to splitting that into a separate file because it can be cached client-side longer - it never needs to change when the rest of your code changes.  It's also total gibberish once minified, but basically it finds the "entry" module function in webpackJsonp and calls it.

So it's the _other_ file that will eventually call that function above (with `"use strict";` in the body).

Reply:
Wonderful. I've removed `setAssertionsEnabled`, removed the tests, and updated the docs.